### PR TITLE
pass options to plugins

### LINKED ISSUE #148: pass options to plugins
follow up for https://github.com/hoodiehq/hoodie/pull/711#issuecomment-306046905

Plugins have currently no way to access options passed to the `Hoodie` constructor. In order to make that possible, I’d suggest we allow to pass another argument to the `hoodie.plugin(plugin)`, so that becomes possible:

```js
hoodie.plugin(function (hoodie, options) {
  hoodie.myMethod = function () {
    // do something with options here
  }
})
```

### PR DESCRIPTION
**Pull request by gr2m**
closes #148

### CODE REVIEW

File: lib/plugin.js [hunk 1]
```diff
@@ -1,15 +1,13 @@
-module.exports = function pluginMethod (plugin) {
-  var self = this
-
+module.exports = function pluginMethod (hoodie, options, plugin) {
   if (typeof plugin === 'function') {
-    plugin(self)
+    plugin(hoodie, options)
```
Comment (mbad0la) [inline lib/plugin.js]: `plugin(hoodie, options)` changes to `plugin(this, options)`

File: lib/plugin.js [hunk 2]
```diff
@@ -1,15 +1,13 @@
-module.exports = function pluginMethod (plugin) {
-  var self = this
-
+module.exports = function pluginMethod (hoodie, options, plugin) {
```
Comment (mbad0la) [inline lib/plugin.js]: `pluginMethod (hoodie, options, plugin)` changes to `pluginMethod (options, plugin)`

File: lib/plugin.js [hunk 3]
```diff
@@ -1,15 +1,13 @@
...
   if (typeof plugin === 'function') {
-    plugin(self)
+    plugin(hoodie, options)
   }
 
   if (typeof plugin === 'object') {
     Object.keys(plugin).forEach(function (key) {
-      self[key] = plugin[key]
+      hoodie[key] = plugin[key]
```
Comment (mbad0la) [inline lib/plugin.js]: How do you plan to pass on `options` when `plugin` is an `object`?

File: tests/specs/plugin.js [hunk 4]
```diff
@@ -33,6 +33,21 @@ test('can define a plugin as a function', function (t) {
...
+    url: 'http://localhost:1234/hoodie',
+    foo: 'bar'
+  })
+
+  hoodie.plugin(function (hoodie, options) {
+    t.is(options.url, 'http://localhost:1234/hoodie')
+    t.is(options.foo, 'bar')
+  })
+})
+
```
Comment (mbad0la) [inline tests/specs/plugin.js]: Need test case for when `hoodie.plugin` takes in an `object` and whether these plugins have access to the `options`

File: lib/get-api.js [hunk 5]
```diff
@@ -54,17 +54,18 @@ function getApi (state) {
...
 
     // events
-    on: require('./events').on.bind(this, state),
-    one: require('./events').one.bind(this, state),
-    off: require('./events').off.bind(this, state),
-    trigger: require('./events').trigger.bind(this, state)
+    on: require('./events').on.bind(null, state),
+    one: require('./events').one.bind(null, state),
+    off: require('./events').off.bind(null, state),
+    trigger: require('./events').trigger.bind(null, state)
```
Comment (gr2m) [inline lib/get-api.js]: it makes no difference. But passing `null` makes clear that the `this` variable is not used inside the methods. We try to omit `this` altogether in all our code bases

File: lib/get-api.js [hunk 6]
```diff
@@ -54,17 +54,18 @@ function getApi (state) {
...
   }
 
+  api.plugin = require('./plugin').bind(null, api, state)
```
Comment (gr2m) [inline lib/get-api.js]: as mentiond above, we try to omit `this` in general. But we need to have access to the `api` that we export as the plugin method must be able to alter it.

File: lib/plugin.js (hunk 3 above)
Comment (gr2m) [inline lib/plugin.js]: we don’t. If you need access to `options` you have to pass in a function, I think that’s fair

File: tests/specs/plugin.js (hunk 4 above)
Comment (gr2m) [inline tests/specs/plugin.js]: see my comment above

**Comment by mbad0la** [comment]
LGTM! 👍

**Comment by gr2m** [comment]
Thanks for the review @mbad0la, these are good questions! As mention several times we are omitting `this` altogether as it’s the root of a lot of headache in the JavaScript world. Instead we are implementing Hoodie in a more functional way by passing in `state` to our api methods. That way it’s also very easy to unit test them :)

**Pull request merged**